### PR TITLE
Added 'EnhancedThemeProvider' control

### LINKED ISSUE #1202: New control [EnhancedThemeProvider]
#### Category
[X] Enhancement
[ ] Bug
[ ] Question

#### Version

Please specify what version of the library you are using: [3.7.2]

#### Expected / Desired Behavior / Question
Improved support for themes and fonts when creating Tab or Personal App in SPFx for Teams or creating Isolated Web Parts

#### Observed Behavior
As specified by issue #1130, the use of Fluent UI React in the version currently used with SPFx 1.14 and earlier, brings with it problems with Teams, and specifically with the High Contrast Theme.

Specifically, if you create a Web Part to be used as a Teams Tab, or as a Personal App, there are two problems:
- Themes provided by the SharePoint page that do not reflect the real Microsoft Teams themes (especially the High Contrast one)
- Lack of default font (Segoe family) when not using Fluent UI React controls.

These two issues frustrate both developers who must write code to implement default font support every time, and users who want to use Web Parts with a dark or high-contrast theme.

I also experienced the problem of fonts when making isolated Web Parts.

#### Steps to Reproduce
Same scenario discussed in issue #1130 

#### Solution
I'm working on creating a control that handles these things:
- Automatic add of correct fonts
- Better support for Teams themes by recreating them in Fluent UI React (Teams supports Fluent UI NorthStar themes)
- Automatisms for automatic check of the Teams theme and use of the corresponding theme.

I will be creating the first PR for this control shortly,
Thanks!

### PR DESCRIPTION
**Pull request by fabiofranzini**
| Q               | A
| --------------- | ---
| Bug fix?        | [ ]
| New feature?    | [X]
| New sample?      | [ ]
| Related issues?  | fixes #1202

#### What's in this Pull Request?

I added the new 'EnhancedThemeProvider' Control used to improve support for themes and fonts when creating Tab or Personal App in SPFx for Teams or creating Isolated Web Parts.

The reasons behind this control are many and concern the use of Fluent UI controls currently officially supported by SPFx, that is:
- `Problems with Teams theme support`, when hosting a Web Part like Tab or Personal App and specifically the lack of support by this version of Fluent UI React of the high contrast theme.
- `Lack of basic style`, such as fonts, for basic HTML elements when creating `Web Parts hosted in Teams as Tabs or personal App`.
- Lack of basic style, such as fonts, for basic HTML elements when creating `Web Parts in "isDomainIsolated" mode`, aka the Isolated Web Parts.

Therefore, the control is to be considered as a sort of `wrapper for all react and non-react controls` that you want to add to the WebPart.

The control `extends the functionality of the Fluent UI ThemeProvider control` (currently in version 7) by adding some logic thanks to the information contained in the 'context' property, that is:
- If the Web Part is hosted inside SharePoint, the theme passed through the 'Theme' property will be used or the default one of the current site will be taken.
- If the web part is hosted within Teams, the "Theme" property will be ignored and using the "Context" property checks which theme is currently applied and adds a handler to notify when the theme is changed. This allows you to manage the change of theme in Teams in real-time, without having to reload the Tab or the Personal App.

As for Teams, given the inconsistency of the theme system of Fluent UI NorthStar (used in Teams) and Fluent UI React (used by SPFx), the themes are "emulated".

The control contains the refining of Teams' `Default`, `Dark` and `Hight Contrast` themes.

This is just an example on Teams with `Hight Contrast` theme applied:
Teams Hight Contrast

### CODE REVIEW

**Comment by fabiofranzini** [comment]
Hi @AJIXuMuK and @joelfmrodrigues with the last commit I finalized the default, dark and high contrast theme for Teams.

As mentioned in the Twitter thread with @AJIXuMuK, I created a repo to individually test the various Fluent controls (not all but most) with the default themes and the ones I created instead.

Specifically, you can see the themes applied using this branch: https://github.com/fabiofranzini/spfx-test-fluent-ui-theme/tree/custom-teams-themes

For this component, as you can see, I have placed the themes in the "common" folder, to share these themes with other components as well (for example the `AdaptiveCardHost` control).

Thanks in advance 🙂

**Comment by joelfmrodrigues** [comment]
@fabiofranzini Many thanks for another great contribution!
I'm just reviewing this now and while testing the solution that you linked above I got one error as it's missing the package @fluentui/react-theme-provider from package.json. Was going to submit a PR, but a bit busy so letting you know here instead as it's a quick fix for you

**Comment by fabiofranzini** [comment]
> @fabiofranzini Many thanks for another great contribution! I'm just reviewing this now and while testing the solution that you linked above I got one error as it's missing the package @fluentui/react-theme-provider from package.json. Was going to submit a PR, but a bit busy so letting you know here instead as it's a quick fix for you

Hi @joelfmrodrigues I just fixed the package.json! Thanks 🙂

File: src/common/fluentUIThemes/FluentUIDefaultTheme.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    const themeColorsFromWindow: any = (window as any).__themeState__.theme;
+    if (themeColorsFromWindow) {
+        currentTheme = createTheme({
+            palette: themeColorsFromWindow
+        });
+    }
+    else
+        currentTheme = getTheme();
+
+    return currentTheme;
```
Comment (joelfmrodrigues) [inline src/common/fluentUIThemes/FluentUIDefaultTheme.ts:14]: This seems to be the same method currently present in `src/controls/adaptiveCardHost/fluentUI/FluentUI.tsx`. Any chance you could double-check and remove the duplication if appropriate?

File: src/controls/EnhancedThemeProvider/EnhancedThemeProvider.tsx [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import * as React from "react";
+import { useCallback, useEffect, useState } from 'react';
+import { fluentUITeamsDarkTheme } from '../../common/fluentUIThemes/FluentUITeamsDarkTheme';
+import { fluentUITeamsDefaultTheme } from '../../common/fluentUIThemes/FluentUITeamsDefaultTheme';
+import { fluentUITeamsHighContrastTheme } from '../../common/fluentUIThemes/FluentUITeamsHighContrastTheme';
+import { IEnhancedThemeProviderProps } from './IEnhancedThemeProviderProps';
+import { ThemeContext, useTheme } from '@fluentui/react-theme-provider';
+import * as telemetry from '../../common/telemetry';
+
+const getDefaultTheme = (): ITheme => {
```
Comment (joelfmrodrigues) [inline src/controls/EnhancedThemeProvider/EnhancedThemeProvider.tsx:14]: Is this the same as above?
image

Can you import it from the common folder?

**Review by joelfmrodrigues — CHANGES_REQUESTED** [review]
@fabiofranzini Great stuff! I just added 2 minor comments as I think you are using the same function in different locations

**Comment by joelfmrodrigues** [comment]
@fabiofranzini sorry just another note. As you are adding a new control, can you also update the documentation links so it get's displayed on the site?
image
No problem if not, I can always do it after. Just thought you were probably not aware of this

File: src/common/fluentUIThemes/FluentUIDefaultTheme.ts (hunk 1 above)
Comment (fabiofranzini) [reply src/common/fluentUIThemes/FluentUIDefaultTheme.ts:14]: Yes, I have a plan to consolidate all this stuff into the common folder... I must do PR for the AdaptiveCardHost.
Basically, this is the first step to consolidate the same themes into the common folder, then I'll use the same code from common folder.

File: src/controls/EnhancedThemeProvider/EnhancedThemeProvider.tsx (hunk 2 above)
Comment (fabiofranzini) [reply src/controls/EnhancedThemeProvider/EnhancedThemeProvider.tsx:14]: Yes! As the previous comment, I'll do a PR (I already made a big refactoring for the AdaptiveCardHost due to share more code from the AdaptiveCardHost and the future AdaptiveCardDesigner controls). If it's ok for you, I prefer to separare changes and have one PF for each controls... Ok for you?

**Comment by fabiofranzini** [comment]
> @fabiofranzini sorry just another note. As you are adding a new control, can you also update the documentation links so it get's displayed on the site? image No problem if not, I can always do it after. Just thought you were probably not aware of this

Great, sorry for that but I've completely missed this step! Thanks again

File: src/controls/EnhancedThemeProvider/EnhancedThemeProvider.tsx (hunk 2 above)
Comment (joelfmrodrigues) [reply src/controls/EnhancedThemeProvider/EnhancedThemeProvider.tsx:14]: Sure, not a problem

**Comment by joelfmrodrigues** [comment]
@fabiofranzini Merged! 👏👏👏

**Comment by fabiofranzini** [comment]
> @fabiofranzini Merged! 👏👏👏

Great, thanks for your feedback! ♥️

**Pull request merged**